homeassistant/components/media_player/plex.py

```diff
@@ -157,99 +286,225 @@ def update_sessions():
+
+        # refresh
+        if session:
+            self._session = session
+        if device:
+            self._device = device
+            self._session = None
+
+        # frozen detection - force state change if past max thresholds
+        frozen_seconds = (round(
+            (datetime.now() - self._last_position_changed).total_seconds()))
+
+        if self.session and self.session.player:
+            if ((frozen_seconds >
+                 int(self.optional_config[CONF_MAX_FROZEN_PAUSED])) and
+                    (self.session.player.state == 'paused' or
+                     self._frozen_fixed_state in (STATE_PAUSED, STATE_IDLE))):
+                _LOGGER.debug('Frozen paused client detected, forcing idle '
+                              'state: %s', self.entity_id)
+                self._frozen_fixed_state = STATE_IDLE
+                self._frozen_content_id = self.media_content_id
+            elif frozen_seconds > int(
+                    self.optional_config[CONF_MAX_FROZEN_PLAYING]
+            ) and self.session.player.state == 'playing':
+                _LOGGER.debug(
+                    'Frozen playing client detected, forcing paused state: %s',
+                    self.entity_id)
+                self._frozen_fixed_state = STATE_PAUSED
+                self._frozen_content_id = self.media_content_id
+            else:
+                self._frozen_fixed_state = None
+
+        if self.session and self.session.player:
+            make = self._convert_na_to_none(self.session.player.device)
+            if make:
+                self._make = make
 
-    def set_device(self, device):
-        """Set the device property."""
-        self.device = device
+    def force_idle(self):
+        """Force client to idle."""
+        self._state = STATE_IDLE
+        self._session = None
 
     @property
     def unique_id(self):
         """Return the id of this plex client."""
-        return '{}.{}'.format(
-            self.__class__, self.device.machineIdentifier or self.device.name)
+        return '{}.{}'.format(self.__class__, self.machine_identifier or
+                              self.name)
 
     @property
     def name(self):
         """Return the name of the device."""
-        return self.device.title or DEVICE_DEFAULT_NAME
+        if self.device:
+            return self._convert_na_to_none(
+                self.device.title) or DEVICE_DEFAULT_NAME
+        if self.session and self.session.player:
+            return self._convert_na_to_none(self.session.player.title)
+
+    @property
+    def machine_identifier(self):
+        """Return the machine identifier of the device."""
+        device_id = None
+        player_id = None
+
+        if self.device:
+            device_id = self._convert_na_to_none(self.device.machineIdentifier)
+
+        if self.session and self.session.player:
+            player_id = self._convert_na_to_none(
+                self.session.player.machineIdentifier)
```

Comment:
Is this doing a network request? We should not do any I/O when fetching properties. Instead, all data should be cached inside the `update` method.